UrbanPop

stevenhofmeyr opened the pull request:
This branch supports the latest UrbanPop data, which includes worker and school assignments. UrbanPop can be used by setting `ic_type=urbanpop`. The  original census implementation (`ic_type=census`) is largely unchanged, but there are some changes that affect both implementations:

1. Removed all calls to `TilingIfNotGPUs`. Inconsistencies in applying this can lead to bugs. Instead, the option `amrex.do_tiling` is now automatically set to 1 for CPUs and 0 for GPUs, and can be overridden on the command-line or in the input file.
2. Moved the initial case settings from `agent.` to `disease.` in the options, since they are specific to diseases. 
3. Added options for neighborhood size and workgroup size (defaults 500 and 20). 
4. Reworked allocations of educators to schools, using a single option for the ratio of students to educators (defaults to 20:1). 
5. Changed the school ID meanings so that an ID of 0 means no school, but otherwise the school ID does not indicate the type of school. Instead, the type of school  is indicated by the school grade of the agent. Also, no longer use -1 for the school ID to indicate closed schools, since that would lose track of the school ID, and instead added a school closed flag.
6. Split the 30-64 age group into two groups: 30-49 and 50 to 64. The latter group has different probabilities of status progression when hospitalized.
7. Added an `examples/inputs.defaults` file which includes all available options, with defaults where they exist.

stevenhofmeyr committed:
Removed demo ic_type.

stevenhofmeyr committed:
Removed unused function

stevenhofmeyr committed:
Made workgroup size and nborhood size into parameters.
Call Redistribute() after moving agents (particles).
Print current infections and deaths counts when running and showing the current day.
Added UrbanPop ic_type and urbanpop_filename parameters.

stevenhofmeyr committed:
Moved the Redistribute call to within the agentMove functions.
Added workgroup_size parameter to read_workerflow

stevenhofmeyr committed:
Moved function definitions from AgentContainer.H to the .cpp file, for all not setter/getter methods.

stevenhofmeyr committed:
Prefixed all AgentContainer variables with m_, as per coding standards

stevenhofmeyr committed:
Moved all functions and data structures that are specific to the census data ic_type into a separate CensusData class, in a separate file. All of the functions previously in Initialization.H/cpp were specific to census data and have thus been moved into the CensusData class.

stevenhofmeyr committed:
Removed geom from CensusData - not needed.

stevenhofmeyr committed:
Merged in development branch updates

stevenhofmeyr committed:
Added python script for processing UrbanPop feather files.
Added header file for UrbanPop data.

stevenhofmeyr committed:
Fixed missing variable in auto-generated UrbanPop header file

stevenhofmeyr committed:
Added UrbanPop data processing files

stevenhofmeyr committed:
Added an option to produce a fixed set of initial infections to try minimize variation between runs.

stevenhofmeyr committed:
Changed work interaction model to bin by workgroup as well as community, which really speeds up CPUs-only runs

stevenhofmeyr committed:
Moved functionality for work interaction model into separate function to conform to same code structure as other interaction models

stevenhofmeyr committed:
Added updated AgentDefinitions file for isAtWork function. Removed unused variables.

stevenhofmeyr committed:
Updated InteractionModel makeBins to always make bins since agents could have been redistributed

stevenhofmeyr committed:
Go through infectious first for home interactions

stevenhofmeyr committed:
For the work interaction model, bin by workgroup plus community. This results in large speedups, especially for CPU only code. Also, when doing interactions, iterate through the infectious agents first, which is also faster, since most agents will be susceptible and fewer will be infectious. This requires atomic multiply for the prob_ptr for GPUs.

stevenhofmeyr committed:
Updated assert calls to avoid unused variables when not in debug mode.

stevenhofmeyr committed:
Changed code to use PTDType for work bins instead of PType, and removed extra AOS attribute from agents

stevenhofmeyr committed:
Use particle tile data for all binning

stevenhofmeyr committed:
Bin home interactions by nborhood and family cluster

stevenhofmeyr committed:
Enable daytime interactions

stevenhofmeyr committed:
Restructured nborhood binning - because of community interactions can't bin smaller than a community, but doing infectious then susceptible still helps performance

stevenhofmeyr committed:
Cleaned up interaction models

stevenhofmeyr committed:
Switched to GPU policy for building bins. Much faster on GPUs

stevenhofmeyr committed:
Implemented new binner for school interactions

stevenhofmeyr committed:
Trying to fix school interactions - should they be in the same nborhood? Work or home?

stevenhofmeyr committed:
Better matching of school interaction model with development branch

stevenhofmeyr committed:
Restructured interaction model to use a common templated function for work, home, school and nborhood. Much simpler code now.

stevenhofmeyr committed:
Always rebuild the bins - needed since redistribute changes the particle locations

stevenhofmeyr committed:
Fixed indentation

stevenhofmeyr committed:
Removed some unused variables, and fixed reference to SOA index for non-group in nborhood interactions.

stevenhofmeyr committed:
Renamed shadowed variables to avoid PR checks complaints.

stevenhofmeyr committed:
Simplified interaction model template by removing the ACT typename.
Switched binning policy from GPU to Serial, in order to preserve reproducibility for a given random seed.
Commented out the generic interaction model, which is never used.

stevenhofmeyr committed:
Merged in development

stevenhofmeyr committed:
Moved makeBins from InteractionModel to InteractionModRandom because that is the only place it is used and it will be replaced soon.

stevenhofmeyr committed:
Added a parameter for the probability of an agent going on random travel

stevenhofmeyr committed:
Removed InteractionModRandom - random travel is now handled in nborhood interactions

stevenhofmeyr committed:
For the interaction models, compute the max group using a parallel reduction, and only do it once.

stevenhofmeyr committed:
Split nborhood interactions into separate models for work and home.
Save the DenseBins for interactions to member of the interaction model. Have an option to reuse without rebuilding. This is disabled at the moment because Redistribute changes the ordering of particles.

stevenhofmeyr committed:
Added command line option agent.fast to switch between fast implementation and slow. Currently this only switches between a serial and a GPU binning policy.

stevenhofmeyr committed:
Removed unused variable

stevenhofmeyr committed:
Fixed issue with double to Real conversion

atmyers committed:
Remove managed memory from read_workerflow

stevenhofmeyr committed:
Fixed overflow when computing number of workgroups for teachers

atmyers committed:
run assignTeachersAndWorkgroup on the host for now

stevenhofmeyr committed:
Merged in fix to avoid using managed memory

stevenhofmeyr committed:
Added option for GPU policy binning to CensusData binning

stevenhofmeyr committed:
Removed some unused code and changed a checking loop to not use managed memory.

stevenhofmeyr committed:
Merged in interactions-bin-by-workgroup branch.

stevenhofmeyr committed:
Added a timer to the main loop output to track how long each time step takes

stevenhofmeyr committed:
Fixes to prevent triggering compiler warnings.

stevenhofmeyr committed:
Simplified binary interactions implementations with shared inline function.

stevenhofmeyr committed:
Further cleaned up interaction probabilities.

stevenhofmeyr committed:
Fixed issue where memory had to be managed

stevenhofmeyr committed:
Ensure that the display of number infected currently agrees with the actual number calculated after updating agents.

stevenhofmeyr committed:
Removed unused variables

stevenhofmeyr committed:
Use ParticleReal to overcome atomic multiply failure in commit check.

stevenhofmeyr committed:
Converted nborhood interaction models to use counting with exponents.

stevenhofmeyr committed:
Modified fast interactions for nborhoods to work on GPUs

stevenhofmeyr committed:
Added a function for getting the community number to reduce code duplication.
Added fast interactions for work.

stevenhofmeyr committed:
Added fast interactions for home

stevenhofmeyr committed:
Implemented alternative fast interactions for neighborhoods, work and home (school is not yet done). These implemenations count the number of infected in each category, and so avoid the n-squared computation.

stevenhofmeyr committed:
Added some comments to the fast work interactions to better explain the computation

stevenhofmeyr committed:
Implemented fast interactions for schools.

stevenhofmeyr committed:
Removed code for slow interactions - that is still available in the interactions-bin-by-workgroup branch.

stevenhofmeyr committed:
Removed unnecessary assert.

stevenhofmeyr committed:
Changed profile function names to differentiate various interaction models.

stevenhofmeyr committed:
Merged in fast-interactions branch

stevenhofmeyr committed:
Working implementation of urbanpop init agents

stevenhofmeyr committed:
Added workgroups to working agents

stevenhofmeyr committed:
Only do the urbanpop agent debug output when using managed memory

stevenhofmeyr committed:
Set the nborhood groups for all the urbanpop agents

stevenhofmeyr committed:
Ensure that families have the same neighborhood

stevenhofmeyr committed:
Added names for profiling functions in UrbanPop initialization. Allow loading of case data for UrbanPop.

stevenhofmeyr committed:
Added vectors and multifabs needed for initial infections to urbanpop class.

stevenhofmeyr committed:
Split infection initialization functions into separate file. Will reuse them for initial infections from urbanpop as well as census data.

stevenhofmeyr committed:
Initialize all components for urbanpop agents init.
Got initial infections working for UrbanPop data, although not perfectly because of community to mesh mismatch.

stevenhofmeyr committed:
Added educators to schools

stevenhofmeyr committed:
Removed need for unit mf from case initialization

stevenhofmeyr committed:
Include NAICS field in agent data and use it to define workgroups

stevenhofmeyr committed:
Merged in development

stevenhofmeyr committed:
Removed assert that called a collective but was only called by one rank

stevenhofmeyr committed:
Added some comments to better understand the columns corresponding to the counts in the output file.

stevenhofmeyr committed:
Subdivide by NAICS for work interactions

stevenhofmeyr committed:
For UrbanPop, use population school grades, not nt_dt_grades.
Set no school to be 0, instead of -1

stevenhofmeyr committed:
Split IntIdx::school into school_grade and school_id, so that in UrbanPop students can be grouped by grade.
Added another attribute, IntIdx::school_closed to represent schools that are closed. The school_id cannot just be set to -1 to indicate closure because the school id needs to be saved for when the school reopens

stevenhofmeyr committed:
Separated the school transmission probabilities from the census-specific definition of school types by school id.

stevenhofmeyr committed:
Allocate teachers to schools by grade. Use grades from worker data.

stevenhofmeyr committed:
Add college interactions for separate transmission

stevenhofmeyr committed:
Assign educators to schools

stevenhofmeyr committed:
Make childcare grouping based on school id, not neighborhood. Create school transmission groups by grade (no impact on Census)

stevenhofmeyr committed:
Setup educators workgroup and nborhood to be determined by school id

stevenhofmeyr committed:
School grade 0 defines daycare/childcare, not the school_id.
Use the neighborhood + 1 for school id, because 0 is no school.
Don't do work interactions for agents at schools.

stevenhofmeyr committed:
Fixed bug for work nborhood interactions where age group 0 was using home nborhoods, even if they had been moved to the work nborhoods

stevenhofmeyr committed:
Use new NAICS codes from updated UrbanPop data.
Draw missing educators from the county level.

stevenhofmeyr committed:
Fixed bugs with new more detailed NAICS.
Made family and school indexes unique only to geoid, to avoid large arrays in interaction models.

stevenhofmeyr committed:
Get global max groups for home interactions to deal with CUDA effectively.
Set max_grid_size to 10 for urbanpop.

stevenhofmeyr committed:
Set correct pos values for x,y locations for urbanopop

stevenhofmeyr committed:
Enabled daytime interactions.

stevenhofmeyr committed:
Restructured urbanpop extract script and changed school ids to be unique to each geoid only, to limit max school id

stevenhofmeyr committed:
Fixed problems with tiling. Now it will set tiling by default if running on CPUs, and not set it if running on GPUs. This works without requiring TilingIfNotGPU() calls to every MFI iteration.
Fixed issue with UrbanPop initialization requiring managed memory - not longer required.
More efficient tracking of communities in interaction models - skip empty grid points to reduce count of communities.

stevenhofmeyr committed:
Fixed school allocations for census. School id of 0 means no school, and workgroups are not set for teachers because they don't interact at work as well as at school.

stevenhofmeyr committed:
Allow workgroup interactions for educators - the workgroup is the school.
Ensure UrbanPop work nborhood is the same as nborhood for non-workers so they can still interact during the day.

stevenhofmeyr committed:
Fix unused variable

stevenhofmeyr committed:
Fixed teacher allocations and some bugs. Added a print out of educator student ratios.

stevenhofmeyr committed:
Moved vector allocation for infected counts in interaction models out of loop to reduce overhead of allocation.

stevenhofmeyr committed:
Removed obsolete timing.

stevenhofmeyr committed:
Merged in changes to terms (latent, incubation)

stevenhofmeyr committed:
Cleaner implementation of lng/lat conversions to and from grid, and also use the correct lng/lat when moving to hospital locations

stevenhofmeyr committed:
Added NM census data and cases

stevenhofmeyr committed:
Added NM UrbanPop data

stevenhofmeyr committed:
Fixed NM inputs.nm paths for data files.

stevenhofmeyr committed:
Merged in development

stevenhofmeyr committed:
Correctly initialize air travel for UrbanPop

stevenhofmeyr committed:
Replace integers with AgeGroup enums where possible.

stevenhofmeyr committed:
Added an additional age bracket.
Restructured hospital treatment and checking for death.

stevenhofmeyr committed:
Got the additional 50-64 age bracket working properly. For census data, assume 57% of previous age group 30to64 are under 50.

stevenhofmeyr committed:
Remove TilingIfNotGPU calls in loops - tiling is set globally once when not using GPUs. Setting it globally ensures there are no bugs caused by tiling in some loops and not others.

stevenhofmeyr committed:
Removed some commented debugging code

stevenhofmeyr committed:
Fixed disease parameters to make it so that the age effect can be an explicit option

stevenhofmeyr committed:
Removed redefinition of variable

stevenhofmeyr committed:
Add print out of age group counts

stevenhofmeyr committed:
Cleaned up options.

stevenhofmeyr committed:
Cleaned up some entries in the documentation

debog committed:
if -> else if so that an agent doesn't enter more than one if-block

debog committed:
renamed isHospitalized() and flag_hosp, etc for clarity

stevenhofmeyr committed:
Moved initial case settings into DiseaseParm structure since they are specific to diseases.

stevenhofmeyr committed:
Renamed enum value of 'random' which was triggering a compilation error

stevenhofmeyr committed:
Merged in development

stevenhofmeyr committed:
Updated outputs: fixed switched Ventilator/ICU label and changed running program stdout details.

debog committed:
making sure latent, incubation, & infection periods are never negative; also ensuring latent period is not greater than the incubation+infection period

debog committed:
simplified the condition for latent period being smaller than incubation+infectious periods

debog committed:
not using std::max since Clang complains!

debog commented on src/AgentDefinitions.H:
```diff
@@ -76,24 +79,41 @@ struct IntIdxDisease
+{
+    enum {
+        none,
+        high_1,         /*!< High School */
+        middle_2,       /*!< Middle School */
+        elem_3,         /*!< Elementary School, nborhoods 1 and 2 */
+        elem_4,         /*!< Elementary School, nborhoods 3 and 4 */
+        day_care_5,     /*!< Day care + Playground */
```
Why did you append the index to the name?

stevenhofmeyr replied:
Because there are actually two sets of enums, one for the school levels in general (`SchoolType`), and one specific to the census data initialization (`SchoolCensusIDType'), which has two levels of elementary school and is only used when setting up the census data. I appended the counter to make it clear when this latter type was being used. Having said that, this is rather ugly and if there is some way of avoiding having this additional census specific enum, that would be a good thing.

debog commented on src/DiseaseStatus.H:
```diff
@@ -168,42 +156,38 @@ void DiseaseStatus<AC,ACT,ACTD,A>::updateAgents(AC& a_agents, /*!< Agent contain
-                                if (timer_ptr[i] > 0) { flag_hosp_ptr[i] = 1; }
-                                if (flag_ICU_i) { flag_ICU_ptr[i] = 1; }
-                                if (flag_vent_i) { flag_vent_ptr[i] = 1; }
+                                if (timer_ptr[i] > 0) flag_hosp_ptr[i] = 1;
```
I think retaining the `{ ... }` even if the `if` condition has only one line is an AMReX practice. Can you check with @atmyers?

stevenhofmeyr replied:
According to the AMReX style [guide](https://github.com/AMReX-Codes/amrex/blob/development/CONTRIBUTING.md#amrex-coding-style-guide), that only applies to `for` loops. I do think it would be a good idea to use something like clang-format with an agreed upon style for consistency across all the contributors. We should discuss as a group.

debog commented on src/DiseaseStatus.H:
```diff
@@ -217,8 +201,8 @@ void DiseaseStatus<AC,ACT,ACTD,A>::updateAgents(AC& a_agents, /*!< Agent contain
                 if (flag_hosp_ptr[i] == 1) {
-                    AMREX_ALWAYS_ASSERT(!isHospitalized(i, ptd));
-                    assign_hospital( i, hosp_i_ptr, hosp_j_ptr, ptd);
+                    // could have already been assignedto a hospital from another disease
+                    if (!isHospitalized(i, ptd)) assign_hospital( i, hosp_i_ptr, hosp_j_ptr, ptd);
```
I don't think the agent has already been assigned to a hospital. We are now outside the `for (int d = 0; d < n_disease; d++ ) {` loop. After it iterates through all diseases, if the `flag_hosp_ptr[i]` is set to `1` (i.e., marked for hospitalization for at least one disease), the agent gets assigned a hospital.

debog commented on src/DiseaseStatus.H:
```diff
@@ -217,8 +201,8 @@ void DiseaseStatus<AC,ACT,ACTD,A>::updateAgents(AC& a_agents, /*!< Agent contain
-                    AMREX_ALWAYS_ASSERT(!isHospitalized(i, ptd));
-                    assign_hospital( i, hosp_i_ptr, hosp_j_ptr, ptd);
+                    // could have already been assignedto a hospital from another disease
+                    if (!isHospitalized(i, ptd)) assign_hospital( i, hosp_i_ptr, hosp_j_ptr, ptd);
```
I don't think the agent has already been assigned to a hospital. We are now outside the `for (int d = 0; d < n_disease; d++ ) {` loop. After it iterates through all diseases, if the `flag_hosp_ptr[i]` is set to `1` (i.e., marked for hospitalization for at least one disease), the agent gets assigned a hospital.

stevenhofmeyr replied:
I put this in there because I was seeing some instances of this - previously there was an assert that an agent was not hospitalized that sometimes failed. Actually, the issue was clearly not about hospitalization from another disease in the same time step, but rather hospitalization in a previous time step. It seems the `counter_ptr` setting should take care of this, but maybe the error you fixed with the infectious/incubation/latent length muddled that up. I mean to say, it was muddled before you fixed the error.

debog replied:
Even without that fix, I don't see how an agent that `AMREX_ALWAYS_ASSERT` would fail, especially since you are probably seeing that fail for a single disease, right? I am staring at that piece of code in both the `development` branch and your `UrbanPop` branch. Given that the incubation period, once sampled from a normal distribution, is fixed for an agent, I don't see why it will enter the `counter_ptr[i] == Math::floor(incubation_period_ptr[i])` block more than once.

debog replied:
This could be a deeper bug. Once my fix is merged, can you put back that assert and see if it fails? If it does, can you share that input file so that I can look into it?

stevenhofmeyr replied:
Yes, likely a deeper bug that I never figured out. I'll merge your fix into UrbanPop and check this again.

stevenhofmeyr replied:
Turns out i was actually testing with 2 diseases when this assertion got triggered. It doesn't for one disease.

debog replied:
Okay, then I definitely need to look into this. Let me know once you are done with all the changes you want to make in this branch, and I'll try it out.

debog replied:
I figured out the reason for this! The logic that I implemented for checking if an agent needs to be hospitalized is not correct for multiple diseases. I inadvertently assumed that all the diseases will have the same incubation period for an agent. So, what you have done here is fine for now, but I need to improve the hospital model.

debog commented on src/HospitalModel.H:
```diff
@@ -132,76 +130,38 @@ void HospitalModel<PCType, PTDType, PType>::treatAgents(PCType& a_agents, /*!< A
-                        // agent is dead
-                        return;
-                    }
+                    // agent is not in hospital
+                    if (!isHospitalized(i, ptd)) return;
+                    // agent just started treatment at start of symptoms appearing
+                    if (counter_ptrs[d][i] == Math::floor(incubation_per_ptrs[d][i])) return;
+                    // agent has recovered/died from disease d
+                    if (timer_ptrs[d][i] == 0) return;
+                    // agent is dead
+                    if (is_alive_ptr[i] == 0) return;
```
`{ return; }` instead of `return;`, unless we are not enforcing that AMReX practice in ExaEpi.

stevenhofmeyr replied:
Not in the style guide.

debog replied:
See this: https://github.com/erf-model/ERF/pull/1410#issuecomment-1919248767

stevenhofmeyr replied:
Ah, ok, I was only aware of that other document, which Andrew pointed me to.

debog commented on src/HospitalModel.H:
```diff
@@ -211,35 +171,49 @@ void HospitalModel<PCType, PTDType, PType>::treatAgents(PCType& a_agents, /*!< A
                         sum_timers += timer_ptrs[d][i];
                     }
-                    if (sum_timers == 0) {
+                    bool is_infected = false;
+                    for (int d = 0; d < n_disease; d++) {
+                        if (status_ptrs[d][i] == Status::infected) {
+                            is_infected = true;
+                            break;
+                        }
+                    }
+                    if (sum_timers == 0 || !is_infected) {
```
Why do we need to check whether the agent is still infected here? If we are at this point in the code, then we considering agents in a hospital whose timers have counted down to the end of hospital duration and are not dead. So, their status must have already been set to `Status::immune` earlier, right?

stevenhofmeyr replied:
Agreed. Fixed.

stevenhofmeyr committed:
Removed unnecessary check for infected in hospital model

debog committed:
incubation period should be less than the latent+infectious periods!

stevenhofmeyr committed:
Merged in fix_disease_status

stevenhofmeyr committed:
Should be incubation period, not latent period.
Check if agent is in hospital before assigning hospital. This can happen in multiple diseases.

stevenhofmeyr committed:
Fixed error in debug build with Real check

debog committed:
fixed the if condition for when hospitalization starts

stevenhofmeyr committed:
Created a function for setting an agent to infected, since that code is used both during updates and when initializing agents to be infected.

stevenhofmeyr committed:
Merge branch 'fix_disease_status' into UrbanPop

stevenhofmeyr committed:
Made sure code is robust to having different precisions for ParticleReal and Real.

stevenhofmeyr committed:
Added comments explaining every option in inputs.defaults

stevenhofmeyr committed:
Formatting and some modified comments

stevenhofmeyr committed:
Changed reduced_inf option to asymp_relative_inf - much clearer.
Automatically set t_hosp_offset to be 3 more than highest hospitalization days, singe this was not an accessible option on the command line.

tannguyen153 commented:
I pulled the PR and ran the code on a single GPU and got this error. Has anyone seen this or it was just me? 
srun -n1 ./agent ../../examples/inputs.nm 
amrex::Abort::0::CUDA error 700 in file /pscratch/sd/n/nnguyent/epi/ExaEpi/build/_deps/amrex-src/Src/Base/AMReX_GpuDevice.cpp line 636: an illegal memory access was encountered !!!
SIGABRT
See Backtrace.0 file for details
MPICH ERROR [Rank 0] [job id 32817036.0] [Mon Nov 11 20:48:29 2024] [nid001136] - Abort(6) (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 6) - process 0

stevenhofmeyr commented:
@tannguyen153 : I haven't seen that error. Can you rerun in debug mode and post a full stack trace?

atmyers committed:
Add a stable mode and make it the default.

atmyers committed:
change default

stevenhofmeyr committed:
Merged in stable redistribute branch

stevenhofmeyr committed:
Merged in development

stevenhofmeyr committed:
Removed strains - this functionality is provided by multiple diseases

stevenhofmeyr committed:
Changed max_grid_size to max_box_size to make it clearer to those not very familiar with amrex terminology.

debog commented on src/AgentContainer.H:
```diff
@@ -108,62 +162,35 @@ public:
-    inline const amrex::Gpu::DeviceVector<int>& getCommElem3TeacherCounts() const {
-        return m_comm_teacher_counts_elem3_d;
-    }
-    inline const amrex::Gpu::DeviceVector<int>& getCommElem4TeacherCounts() const {
-        return m_comm_teacher_counts_elem4_d;
-    }
-    inline const amrex::Gpu::DeviceVector<int>& getCommTeacherCounts() const {
-        return m_comm_teacher_counts_total_d;
-    }
-
-    /*! \brief return the symptomatic withdrawal status */
-    inline int symptomaticWithdraw() const {
-        return m_symptomatic_withdraw;
-    }
-
     /*! \brief return the symptomatic withdrawal compliance */
     inline amrex::Real symptomaticWithdrawCompliance() const {
         return m_symptomatic_withdraw_compliance;
     }
 
+    void printStudentTeacherCounts() const;
+
+    void printAgeGroupCounts() const;
+
     amrex::iMultiFab m_student_counts;
-    amrex::Gpu::DeviceVector<int> m_unit_teacher_counts_d;
-    amrex::Gpu::DeviceVector<int> m_comm_teacher_counts_total_d;
-    amrex::Gpu::DeviceVector<int> m_comm_teacher_counts_high_d;
-    amrex::Gpu::DeviceVector<int> m_comm_teacher_counts_middle_d;
-    amrex::Gpu::DeviceVector<int> m_comm_teacher_counts_elem3_d;
-    amrex::Gpu::DeviceVector<int> m_comm_teacher_counts_elem4_d;
-    amrex::Gpu::DeviceVector<int> m_comm_teacher_counts_daycr_d;
-    std::vector<long> m_student_teacher_ratios = {20, 20, 20, 20, 20, 1000000000}; // ignore daycare teachers since daycare interaction isnt handled yet!
+    amrex::Real m_student_teacher_ratio = 15; /*!< Used only for Census data */
```
The teacher-to-student ratio may differ between the various school grades. Is it possible to retain school-type-dependent ratios?

stevenhofmeyr replied:
Yes of course. I'm not sure exactly why I removed that....

stevenhofmeyr committed:
Made the school ratio option used for census data into an array of values, one per school type

stevenhofmeyr committed:
Fixed new bug in teacher count initialization

stevenhofmeyr committed:
Default to using managed memory as that is required for the serial binning, which is required to get consistent results for a given random seed and so is set by default.

atmyers commented:
My understanding is there may be a bug with this (@tannguyen153's issue) that only affects runs with UrbanPop, and that it also makes a number of other changes that others are waiting on. So I'd propose going ahead and merging this, and @tannguyen153's issue can be addressed in a follow-on. Does that sound good to you @stevenhofmeyr?

tannguyen153 commented:
Sounds good to me. I can just pull the development branch to test UrbanPop features. I'll have more time to run tests tomorrow after finishing some work for SC24 today.

stevenhofmeyr commented:
Ok, thanks @tannguyen153  and @atmyers. Let's merge the PR. @tannguyen153 : if you can let me know what the conditions are to reproduce the bug, I can set about fixing it.
